### PR TITLE
Feature: Fast search

### PR DESCRIPTION
**Pull request by parasharrajat**
### Details

https://github.com/Expensify/App/issues/6069#issuecomment-955887403

### Fixed Issues

$ https://github.com/Expensify/App/issues/6069

### Tests | QA Steps
1. Open newDot on web or desktop.
2. Used Shortcut keys to open the search page.
3. type something immediately after.
4. When the search page is shown, your search input should not be lost.
5. and you should see searched results based on the input.

### Tested On

- [x] Web
- [x] Mobile Web (Not affected)
- [x] Desktop
- [x] iOS (Not affected)
- [x] Android (Not affected)

#### Web | Desktop

https://user-images.githubusercontent.com/24370807/141067794-08c8f867-f35a-404e-8117-6ca55b58837f.mp4

#### Android

Screenshot 2021-11-10 12:54:56

### CODE REVIEW

**Comment by parasharrajat** [comment]
Please let me know if that works for you.

There is one point worth mentioning that I am feeling to talk about. I think the current job value is not worth the efforts I have put together into this. I am following this issue since the start and also helped in the previous attempt(PR) to fix this. If you think that this is justified, then I request for a price increase.

But if you think that price is sufficient for this job, then it's fine.

File: src/libs/KeyboardShortcut/index.js [hunk 1]
```diff
@@ -124,6 +126,56 @@ function subscribe(key, callback, modifiers = 'shift', captureOnInputs = false)
     return () => unsubscribe(key);
 }
 
+
```
Comment (Luke9389) [inline src/libs/KeyboardShortcut/index.js:129]: there's an extra line of whitespace here.

File: src/libs/KeyboardShortcut/index.js [hunk 2]
```diff
@@ -124,6 +126,56 @@ function subscribe(key, callback, modifiers = 'shift', captureOnInputs = false)
...
+/**
+ * Captures all the key inputs until explictly consumed.
+ *
+ * @param {boolean} [captureOnInputs=false] Should we capture the event on inputs too?
```
Comment (Luke9389) [inline src/libs/KeyboardShortcut/index.js:133]: I'm not entirely sure what this means. Could you give me an example of when we would want this to be `true`?

File: src/libs/KeyboardShortcut/index.js [hunk 3]
```diff
@@ -124,6 +126,56 @@ function subscribe(key, callback, modifiers = 'shift', captureOnInputs = false)
...
+    if (keyBufferCallback) {
+        return;
+    }
+    keyBufferCallback = (evt) => {
+        if (!captureOnInputs
+            && evt.target.nodeName === 'INPUT'
+            && evt.target.nodeName === 'TEXTAREA'
+            && evt.target.contentEditable === 'true') {
+            return;
+        }
```
Comment (Luke9389) [inline src/libs/KeyboardShortcut/index.js:145]: Does this mean we are going to early return if the user is typing in an input box normally?

File: src/pages/SearchPage.js [hunk 4]
```diff
@@ -159,31 +161,32 @@ class SearchPage extends Component {
...
-            <ScreenWrapper>
-                {({didScreenTransitionEnd}) => (
-                    <>
-                        <HeaderWithCloseButton
-                            title={this.props.translate('common.search')}
-                            onCloseButtonPress={() => Navigation.dismissModal(true)}
+            <ScreenWrapper
+                onTransitionEnd={() => {
+                    const searchValue = KeyboardShortcut.consumeBuffer();
+                    this.setState({didScreenTransitionEnd: true, searchValue}, this.updateOptions);
```
Comment (Luke9389) [inline src/pages/SearchPage.js:167]: I think this looks good. 👍 It's cool to see it in action.

File: src/components/ScreenWrapper.js [hunk 5]
```diff
@@ -72,6 +72,7 @@ class ScreenWrapper extends React.Component {
             }
 
             Navigation.dismissModal();
+            KeyboardShortcut.consumeBuffer();
```
Comment (Luke9389) [inline src/components/ScreenWrapper.js:75]: This may be obvious, but why do we have this here?

**Review by Luke9389 — CHANGES_REQUESTED** [review]
In general, I think this is a handy solution that will likely be applied elsewhere in the app (as we continue to add features with more text inputs with transitions. However, I do think the best route is to minimize the number of components we add this to. 

I left some questions below.

File: src/libs/KeyboardShortcut/index.js (hunk 1 above)
Comment (parasharrajat) [reply src/libs/KeyboardShortcut/index.js:129]: Oops

File: src/libs/KeyboardShortcut/index.js (hunk 2 above)
Comment (parasharrajat) [reply src/libs/KeyboardShortcut/index.js:133]: I just picked it up from the old Keyboard handlers on the file but I think it better to just remove it for now. Right?

File: src/libs/KeyboardShortcut/index.js (hunk 3 above)
Comment (parasharrajat) [reply src/libs/KeyboardShortcut/index.js:145]: Yeah. This is supposed to work in that way. If the user is typing in the focused input then don't buffer the value. Useful in cases where we want to use the buffer in an input that will be autofocused at some moment in time. So to minimize duplicate

File: src/pages/SearchPage.js (hunk 4 above)
Comment (parasharrajat) [reply src/pages/SearchPage.js:167]: This does not only put value in the input. It keeps the searched list ready to be shown to the user based on his input. So he does not have to wait for the search to happen.

File: src/components/ScreenWrapper.js (hunk 5 above)
Comment (parasharrajat) [reply src/components/ScreenWrapper.js:75]: For example, a user triggers the search page and this is not yet mounted and he immediately presses ESC to close the page. This is a fail-safe to clear the buffer.

**Comment by parasharrajat** [comment]
@marcaaron I am waiting for your review here, it won't move forward without it. Please review it in your free time.

**Comment by marcaaron** [comment]
@parasharrajat I've been discussing alternatives with @Luke9389 1:1. I feel pretty strongly that the proposed solution is unnecessary. More details soon.

**Comment by Luke9389** [comment]
Alright, @parasharrajat. We're not going to move forward with this solution, but we do have an alternative for you. 

If you'd be willing to check out [this diff](https://github.com/Expensify/App/compare/marcaaron-captureInput?expand=1), it shows an alternative solution path that doesn't _completely_ remove the behavior of losing keystrokes, but does alleviate it somewhat, and is a super low impact solution.

**Comment by parasharrajat** [comment]
That was my first technique to resolve it. But it won't work if want to autofocus the input during animation.
Autofocus breaks animation so you can't focus input until the animation is done thus can't type (original issue).

**Comment by marcaaron** [comment]
>Autofocus breaks animation so you can't focus input until the animation is done thus can't type (original issue).

Can we resolve the underlying issue instead of adding cruft? Can you remind us why autofocusing the input breaks animation? Maybe we should look at that issue again...

**Comment by parasharrajat** [comment]
> Can you remind us why autofocusing the input breaks animation?
I don't know the exact reason but we have faced this issue several times where we decided to focus after the animation is done. For eg, using InteractionManager. 

The browser will try to scroll the input in the view when input is focused which somehow causes the animations to become janky as well as behave abruptly. On Safari mobile Web,  the sideModal screens showed jumping behavior.

**Comment by marcaaron** [comment]
Feels like our new goal should be to prevent jank when focusing the input then? 

I may be alone here but I feel we are just adding a workaround on top of a workaround. 

I'm aware that I'm delaying this fix and appreciate the solutions and discussion. But this issue is not urgent and the tricks and workarounds at some point need to be improved upon. Some workarounds are fine for the first iteration or proof of concept for a project, but they should not be confused with long term solutions.

I want to propose we close this PR. Add the context to the issue detailing what needs to be done. And try to figure out the true cause of the animation jank and address it properly. Of course, we will pay for the investigation - but I don't think this PR is on the right path. 

Is everyone good with that?

**Comment by parasharrajat** [comment]
I don't have any issues with it but still feel that there is no better way of fixing the jank animation due to focus. I would be excited to see a solution.

**Comment by Luke9389** [comment]
Ok, I am also in agreement that we should close this, post some details and restate our objectives on the issue.

**Comment by Luke9389** [comment]
@parasharrajat I want to thank you for your willingness to table that solution in service of keeping our codebase simple and intuitive. It's clear that you're willing to think outside the box, which is really cool to see.

**Pull request closed without merging**